### PR TITLE
Issue 18354 more normal

### PR DESCRIPTION
**Pull request by wezell**

### CODE REVIEW

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java [hunk 1]
```diff
@@ -36,7 +36,12 @@ public String getRequestURI() {
                  2. A ".." segment is removed only if it is preceded by a non-".." segment.
                  3. Normalization has no effect upon opaque URIs. (mailto:a@b.com)
                  */
-                return URI.create(super.getRequestURI()).normalize().toString();
+                String newNormal = URI.create(super.getRequestURI()).normalize().toString();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java:39]: I think you could use something like
```
return newNormal.replaceAll(Pattern.quote("//"), StringPool.SLASH);
```

File: dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java [hunk 2]
```diff
@@ -36,7 +36,12 @@ public String getRequestURI() {
...
+                
+                while(newNormal.indexOf("//")>-1) {
+                    newNormal = newNormal.replace("//", "/");
+                }
+                return newNormal;
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java:44]: miss test here

**Review by freddyucv — CHANGES_REQUESTED** [review]

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java (hunk 1 above)
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java:39]: While the code looks "cleaner", my thinking was that a replaceAll with a RegEx was actually going to be slower than just doing a string replace in a loop.

File: dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java [hunk 3]
```diff
@@ -111,5 +135,73 @@ private void compare(final String originalURI, final String expectedNormalizedUR
         assertNotNull(normalizedValueByFilter);
         assertEquals(expectedNormalizedURI, normalizedValueByFilter);
     }
+    
+    /**
+     * Test to verify the {@link NormalizationFilter} is applying properly the normalization on
+     * invalids URIs
+     */
+    @Test
+    public void test_uri_normalizer_fixes_double_slashes() throws IOException, ServletException {
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java:144]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java [hunk 4]
```diff
@@ -86,15 +109,16 @@ public void test_uri_normalization_valid_URI() throws IOException, ServletExcept
         String originalURI = "/folder/important/secret_file.dat";
         compare(originalURI, originalURI);
 
+        // (query strings are not part of URI)
         originalURI = "/folder/folder1/forward_jsp.jsp?FORWARD_URL=http://google.com";
-        compare(originalURI, originalURI);
+        compare(originalURI, originalURI.substring(0,originalURI.indexOf("?")));
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java:114]: Codacy Issue found: [String.indexOf(char) is faster than String.indexOf(String).](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java [hunk 5]
```diff
@@ -62,18 +63,40 @@ public void test_uri_normalization_invalid_URI() throws IOException, ServletExce
 
         // Each "." segment is simply removed
         originalURI = "./folder/folder1/file.dat";
-        expectedNormalizedURI = "folder/folder1/file.dat";
+        expectedNormalizedURI = "/folder/folder1/file.dat";
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java:66]: Codacy Issue found: [The String literal "/folder/folder1/file.dat" appears 4 times in this file; the first occurrence is on line 66](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java [hunk 6]
```diff
@@ -30,13 +37,35 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
...
+                    return newNormal;
+                }
+                catch(IllegalArgumentException ill) {
+                    Logger.warnAndDebug(this.getClass(),ill);
+                    HttpServletResponse response= (HttpServletResponse) servletResponse;
+                    try {
+                        response.sendError(HttpServletResponse.SC_FORBIDDEN);
+                        response.flushBuffer();
+                    }
+                    catch(Exception e) {}
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java:65]: Codacy Issue found: [Avoid empty catch blocks](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java [hunk 7]
```diff
@@ -111,5 +135,73 @@ private void compare(final String originalURI, final String expectedNormalizedUR
...
+        compare(originalURI, expectedNormalizedURI);
+    }
+    
+    
+    /**
+     * Test to verify the {@link NormalizationFilter} is applying properly the normalization on
+     * invalids URIs
+     */
+    @Test
+    public void test_uri_normalizer_invalid_uris() throws IOException, ServletException {
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java:198]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java [hunk 8]
```diff
@@ -111,5 +135,73 @@ private void compare(final String originalURI, final String expectedNormalizedUR
...
+        // testing escaped slashes - Filter barfs and returns /
+        String originalURI = "/\\///folder//important////secret_file.dat";
```
Comment (dev-dotcms) [inline dotCMS/src/test/java/com/dotcms/filters/NormalizationFilterTest.java:200]: Codacy Issue found: [Local variable 'originalURI' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

File: dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/filters/NormalizationFilter.java:65]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5436097)

**Pull request merged**